### PR TITLE
[FLINK-17632][yarn] Support to specify a remote path for job jar

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change

After FLINK-13938, we could support to register remote shared libs as Yarn local resources and prevent unnecessary uploading and downloading for system Flink jars.

However, we still need to specify a local user jar to run a Flink job on Yarn. This ticket aims to add the remote path support. It have at least two purposes.

* Accelerate the submission
* Better integration with application mode. Since the user main code is executed in the cluster(jobmanager), we do not need the user jar exists locally.
A very typical use case is like following.

```
./bin/flink run-application -p 10 -t yarn-application \
-yD yarn.provided.lib.dirs="hdfs://myhdfs/flink/lib" \
hdfs://myhdfs/jars/WindowJoin.jar
```

## Brief change log

* `a251fc` Always build the packaged program in the cluster for application mode
* `9c3e41` Support to specify a remote path for job jar
* `f257f9` Move some variables as member of YarnApplicationFileUploader to make it more self-contained

## Verifying this change

* New added unit tests
* ITCase for Yarn application mode
* Manually test in a real Yarn cluster

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (will add document in a separate ticket with all other application features)

### CODE REVIEW

**Comment by wangyang0918** [comment]
cc @kl0u since it is part of FLIP-85(application mode). Could you help with review this PR.

**Comment by wangyang0918** [comment]
@flinkbot run azure

**Comment by wangyang0918** [comment]
@kl0u The pre-check in the getRemoteSharedPaths  is a good idea. It could make the submission fails fast if the yarn.provided.lib.dirs is wrongly configured to local directories.

Thanks a lot for your review. I have addressed all the comments in your branch and force-pushed. Please have another look.

**Comment by wangyang0918** [comment]
rebase latest master

File: flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java [hunk 1]
```diff
@@ -733,67 +737,49 @@ private ApplicationReport startAppMaster(
...
-		fileReplication = fileReplication > 0 ? fileReplication : yarnFileReplication;
-
 		// only for per job mode
 		if (jobGraph != null) {
 			for (Map.Entry<String, DistributedCache.DistributedCacheEntry> entry : jobGraph.getUserArtifacts().entrySet()) {
-				org.apache.flink.core.fs.Path path = new org.apache.flink.core.fs.Path(entry.getValue().filePath);
 				// only upload local files
-				if (!path.getFileSystem().isDistributedFS()) {
-					Path localPath = new Path(path.getPath());
+				if (Utils.isRemotePath(entry.getValue().filePath)) {
```
Comment (klion26) [inline flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:754]: Here the comment said that `only upload local files`, but the logic in `if` will be true only for `RemotePath`, so maybe this is a bug? @wangyang0918

Comment (wangyang0918) [reply flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java:754]: @klion26 Thanks for your comments about this. It is indeed a bug and should be fixed in 1.11. I will create a ticket right now. Moreover, i will dig in and find why the related tests does not fail.

**Pull request closed without merging**